File: MyPrototypes/PsCore/PsDataCoreVb/SqlDataCore.vb
Imports System.Text.RegularExpressions
Imports System.Data.SqlClient
Imports System.Configuration
Imports System.Data.Common
'Namespace Ps.Datacore
'    Public Class SqlDataCore
'        ''' <summary>
'        ''' Propiedad privada que sirve para realizar la conexión a la base de datos de Oracle.
'        ''' </summary>
'        ''' <remarks></remarks>
'        Private ReadOnly connection As SqlConnection
'        ''' <summary>
'        ''' Propiedad privada que sirve para la creación y manejo de comandos de Oracle.
'        ''' </summary>
'        ''' <remarks></remarks>
'        Private command As SqlCommand
'        ''' <summary>
'        ''' Propiedad privada que sirve como contenedor de los datos extraídos de la base de datos Oracle.
'        ''' </summary>
'        ''' <remarks></remarks>
'        Private ds As DataSet
'        ''' <summary>
'        ''' Propiedad privada que sirve de enlace entre el contenedor de los datos y la base de datos Oracle.
'        ''' </summary>
'        ''' <remarks></remarks>
'        Private da As SqlDataAdapter
'        ''' <summary>
'        ''' Constructor de la clase que sirve para la inicialización de ciertos valores requeridos al momento de realizar el acceso a la base de datos Oracle.
'        ''' </summary>
'        ''' <remarks></remarks>
'        Public Sub New()
'            Dim cs = ConfigurationManager.ConnectionStrings("PsConnection")
'            Dim str As String
'            If cs Is Nothing Then
'                '´´str = Decifrar(ConfigurationManager.ConnectionStrings("cadenaConexionProd").ConnectionString)
'                str = ConfigurationManager.ConnectionStrings("PsConnectionProd").ConnectionString
'            Else
'                str = cs.ConnectionString
'            End If

'            Me.connection = New SqlConnection(str)
'        End Sub
'        Public Property Transaction() As SqlTransaction
'            Get
'                Return m_Transaction
'            End Get
'            Set(value As SqlTransaction)
'                m_Transaction = value
'            End Set
'        End Property
'        Private m_Transaction As SqlTransaction
'        ''' <summary>
'        ''' Método privado que permite la apertura de la conexión a la base de datos Oracle.
'        ''' </summary>
'        ''' <remarks></remarks>
'        Public Sub OpenConnection()
'            If connection.State = ConnectionState.Closed Then
'                Me.connection.Open()
'            End If

'        End Sub
'        ''' <summary>
'        ''' Método que extrae datos de la base de datos Oracle basado en un procedimiento almacenado (Stored Procedure) y los parámetros que este recibe.
'        ''' </summary>
'        ''' <param name="spName">Nombre del procedimiento almacenado (Stored Procedure)</param>
'        ''' <param name="spParams">Arreglo de parámetros requeridos por el procedimiento almacenado (Stored Procedure)</param>
'        ''' <returns>Retorna un DataSet con los datos obtenidos de la base de datos Oracle extraídos mediante la ejecución del procedimiento almacenado (Stored Procedure)</returns>
'        ''' <remarks></remarks>
'        Public Function GetDataBySP(ByVal spName As String, ByVal ParamArray spParams() As SqlParameter) As DataSet
'            ' Inicialización de la propiedad propiedad privada que servirá de contendor de datos
'            Me.ds = New DataSet()

'            IniCommandAddParam(spName, spParams)
'            ' Se crea el adaptador que servirá de enlace entre la base de datos y el contenedor de datos
'            Me._da = New SqlDataAdapter(Me.command)
'            Try
'                Me.OpenConnection()
'                ' Se cargan los datos de la base de datos
'                Me._da.Fill(Me.ds)
'                ' Se cierra la conexión a la base de datos

'            Catch ex As Exception
'                Throw New Exception(ex.Message)
'            Finally
'                Me.CloseConnection()
'            End Try

'            Return Me.ds
'        End Function
'        Private Sub IniCommandAddParam(ByVal spName As String, ByVal ParamArray spParams() As SqlParameter)
'            Me.command = New SqlCommand(spName, Me.connection)
'            Me.command.CommandType = commandType.StoredProcedure
'            For Each spParam As SqlParameter In spParams
'                Me.command.Parameters.Add(spParam)
'            Next
'        End Sub
'        Friend Sub ExecuteCommandScalar(ByVal spName As String, ByVal ParamArray spParams() As SqlParameter)
'            IniCommandAddParam(spName, spParams)
'            Try
'                Me.OpenConnection()
'                Me.command.ExecuteScalar()
'            Catch ex As Exception
'                Throw New Exception(ex.Message)
'            Finally
'                Me.CloseConnection()
'            End Try
'        End Sub
'        Public Sub ExecuteCommandScalarTrans(ByVal spName As String, ByVal ParamArray spParams() As SqlParameter)
'            IniCommandAddParam(spName, spParams)
'            Try
'                Me.command.ExecuteScalar()
'            Catch ex As Exception
'                Throw New Exception(ex.Message)
'            Finally
'                Me.CloseConnection()
'            End Try
'        End Sub
'        Public Sub RollbackTrans()
'            command.Transaction.Rollback()
'        End Sub
'        Public Sub BeginTrans()
'            OpenConnection()
'            command.Transaction = connection.BeginTransaction
'        End Sub
'        Public Sub CommitTrans()
'            command.Transaction.Commit()
'        End Sub
'        ''' <summary>
'        ''' Método que permite la liberación de memoria de los objetos utilizados en el acceso a datos.
'        ''' </summary>
'        ''' <remarks></remarks>
'        Friend Sub Dispose()
'            Me.ds.Dispose()
'            Me.da.Dispose()
'            Me.command.Dispose()
'            If Not Me.connection.State = ConnectionState.Closed Then
'                Me.CloseConnection()
'            End If
'        End Sub
'        Private _da As SqlDataAdapter
'        Public Property Command() As SqlCommand
'            Get
'                Return m_Command
'            End Get
'            Set(value As SqlCommand)
'                m_Command = value
'            End Set
'        End Property
'        Private m_Command As SqlCommand
'        Public Property commandType() As CommandType
'            Get
'                Return m_commandType
'            End Get
'            Set(value As CommandType)
'                m_commandType = value
'            End Set
'        End Property
'        Private m_commandType As CommandType
'        'Public Property Ds() As DataSet
'        '    Get
'        '        Return m_Ds
'        '    End Get
'        '    Set(value As DataSet)
'        '        m_Ds = value
'        '    End Set
'        'End Property
'        Private m_Ds As DataSet
'        ''' <summary>
'        ''' Método privado que sirve para cerrar la conexión a la base de datos de MSSQL.
'        ''' </summary>
'        ''' <remarks></remarks>
'        Public Sub CloseConnection()
'            If connection.State = ConnectionState.Open Then
'                Me.connection.Close()
'            End If
'        End Sub
'        Public Function GetDataIdSet(ByVal str As String) As SqlDataAdapter
'            GetDataIdSet = New SqlDataAdapter
'            Me.OpenConnection()
'            GetDataIdSet = New SqlDataAdapter(str, connection)
'            Me.OpenConnection()
'        End Function
'        Public Function FillCommand(sqlCommand As String, values As [Object](), Optional paramDirs As [Object]() = Nothing) As DbCommand
'            Me.LoadCommandObj(sqlCommand, values, paramDirs)
'            Return command
'        End Function
'        Public Function FillCommand(procedureName As String, values As [Object](), parameterNames As [String], Optional paramDirs As [Object]() = Nothing) As DbCommand
'            Me.LoadCommandObj(procedureName, values, paramDirs, True, parameterNames)
'            Return command
'        End Function
'        'Dim isTransaction As Boolean = False
'        Private Sub LoadCommandObj(sqlCommand As [String], values As [Object](), Optional paramDirs As [Object]() = Nothing, Optional isProcedure As Boolean = False, Optional parameters As String = Nothing)
'            command = New SqlCommand()
'            command.Connection = connection
'            If isTransaction Then
'                command.Transaction = Transaction
'            End If
'            command.CommandText = sqlCommand
'            Dim names As New ArrayList()
'            If isProcedure Then
'                Dim prms As String() = parameters.Split(","c).[Select](Function(r) r.Trim()).ToArray()
'                names.AddRange(prms)
'                command.CommandType = commandType.StoredProcedure
'            Else
'                names = GetParameterNames(sqlCommand)
'                command.CommandType = commandType
'            End If
'            Dim index As Integer = 0
'            If values IsNot Nothing Then
'                For Each val As [Object] In values
'                    If paramDirs Is Nothing Then
'                        If names.Count = 0 Then
'                            AddWithValue("", val)
'                        Else
'                            AddWithValue(names(index).ToString(), val)
'                        End If
'                    Else
'                        If paramDirs.Contains(names(index)) Then
'                            AddWithValue(names(index).ToString(), val, ParameterDirection.Output)
'                        ElseIf Not paramDirs.Contains(names(index)) AndAlso paramDirs.Length < values.Length Then
'                            AddWithValue(names(index).ToString(), val)
'                        Else
'                            AddWithValue(names(index).ToString(), val, DirectCast([Enum].Parse(GetType(ParameterDirection), paramDirs(index).ToString()), ParameterDirection))
'                        End If
'                    End If
'                    index += 1
'                Next
'            End If
'        End Sub
'        Public Function GetParameterNames(query As String) As ArrayList
'            Dim paramNames As New ArrayList()
'            Dim pattern As New Regex(":([A-Za-z0-9]+)(\s*)")
'            For Each match As Match In pattern.Matches(query)
'                paramNames.Add(match.Value.Replace(":"c, " "c).Trim())
'            Next
'            Return paramNames
'        End Function
'        Public Sub AddWithValue(paramName As String, value As Object, Optional direction As ParameterDirection = ParameterDirection.Input)
'            Dim param = command.CreateParameter()
'            param.ParameterName = paramName
'            param.Value = value
'            param.Direction = direction
'            command.Parameters.Add(param)
'        End Sub
'    End Class
'End Namespace

File: MyPrototypes/PsCore/PsDataCore/SqlDataCore.cs
using System;
using System.Collections;
using System.Configuration;
using System.Data;
using System.Data.Common;
using System.Data.SqlClient;
using System.Linq;
using System.Text.RegularExpressions;

namespace PsDataCore
{
    public class SqlDataCore
    {
        private readonly SqlConnection _connection;
        private SqlCommand _command;
        private DataSet _ds;
        private SqlDataAdapter _da;

        public SqlCommand Command { get; set; }
        public SqlTransaction Transaction { get; set; }
        private bool isTransaction;
        public CommandType commandType { get; set; }
        public DataSet Ds { get; set; }
        public SqlDataCore()
        {
            //iniciando el constructor con el tipo que usare mas amenudo
            commandType = CommandType.StoredProcedure;
            var cs = ConfigurationManager.ConnectionStrings["PsConnection"];

            //string _str = cs == null ? Decifrar(ConfigurationManager.ConnectionStrings["cadenaConexionProd"].ConnectionString) : cs.ConnectionString;
            string _str = cs == null ? (ConfigurationManager.ConnectionStrings["PsConnectionProd"].ConnectionString) : cs.ConnectionString;
            this._connection = new SqlConnection(_str);
        }

        public static void VerifyConexion()
        {
            var cs = ConfigurationManager.ConnectionStrings["PsConnection"];
            //string _str = cs == null ? Decifrar(ConfigurationManager.ConnectionStrings["cadenaConexionProd"].ConnectionString) : cs.ConnectionString;
            string _str = cs == null ? (ConfigurationManager.ConnectionStrings["PsConnectionProd"].ConnectionString) : cs.ConnectionString;
            var _connection = new SqlConnection(_str);
        }

        public void OpenConnection()
        {
            this._connection.Open();
        }


        public DataSet GetDataBySP(string spName, params SqlParameter[] spParams)
        {

            this._ds = new DataSet();

            IniCommandAddParam(spName, spParams);

            this._da = new SqlDataAdapter(this._command);
            try
            {
                this.OpenConnection();
                this._da.Fill(this._ds);

            }
            catch (Exception ex)
            {
                throw new Exception(ex.Message);
            } finally
            {
                this.CloseConnection();
            }

            return this._ds;
        }

        private void IniCommandAddParam(string spName, params SqlParameter[] spParams)
        {
            this._command = new SqlCommand(spName, this._connection);
            this._command.CommandType = CommandType.StoredProcedure;

            foreach (var param in spParams)
            {
                this._command.Parameters.Add(param);
            }
        }


        public void ExecuteCommandScalar(string spName, params SqlParameter[] spParams)
        {
            IniCommandAddParam(spName, spParams);
            try
            {
                this.OpenConnection();
                this._command.ExecuteScalar();
            }
            catch (Exception ex)
            {
                throw new Exception(ex.Message);
            } finally
            {
                this.CloseConnection();
            }

        }

        protected void Dispose()
        {
            this._ds.Dispose();
            this._da.Dispose();
            this._command.Dispose();
            if (this._connection.State != ConnectionState.Closed)
            {
                this.CloseConnection();
            }
        }

        public void CloseConnection()
        {
            this._connection.Close();
        }

        //private static string Decifrar(string pCifrado)
        //{
        //    var textConverter = new ASCIIEncoding();
        //    var myRijndael = new System.Security.Cryptography.RijndaelManaged();
        //    byte[] fromEncrypt = null;
        //    byte[] encrypted = Convert.FromBase64String(pCifrado.Replace(";", "ON").Replace(",", "on").Replace(".", "+"));
        //    byte[] key =
        //    {
        //        12,
        //        215,
        //        88,
        //        34,
        //        53,
        //        68,
        //        100,
        //        183,
        //        95,
        //        132,
        //        56,
        //        74,
        //        46,
        //        27,
        //        39,
        //        62
        //    };
        //    byte[] IV =
        //    {
        //        88,
        //        234,
        //        153,
        //        46,
        //        27,
        //        39,
        //        12,
        //        215,
        //        62,
        //        68,
        //        10,
        //        183,
        //        44,
        //        132,
        //        56,
        //        64
        //    };
        //    System.Security.Cryptography.ICryptoTransform decryptor = myRijndael.CreateDecryptor(key, IV);
        //    var msDecrypt = new System.IO.MemoryStream(encrypted);
        //    var csDecrypt =
        //        new System.Security.Cryptography.CryptoStream(msDecrypt, decryptor,
        //            System.Security.Cryptography.CryptoStreamMode.Read);

        //    fromEncrypt = new byte[encrypted.Length + 1];
        //    csDecrypt.Read(fromEncrypt, 0, fromEncrypt.Length);
        //    return textConverter.GetString(fromEncrypt);
        //}






        /// <summary>
        /// Propiedad publica de solo lectura que especifica si para la
        /// conexión actual hacia la BD, se esta utilizando una transacción
        /// explícita.
        /// </summary>
        public bool IsTransaction
        {
            get { return isTransaction; }
        }

        /// <summary>
        /// Establece para el ámbito de conexión actual hacia la base de
        /// datos, se utilizará un manejo explícito de la transacción 
        /// asociada a las consultas
        /// </summary>
        public void BeginTransaction()
        {
            if (!isTransaction && _connection.State == ConnectionState.Open)
            {
                Transaction = _connection.BeginTransaction();
                isTransaction = true;
            }
        }
        public void RollBackTransaction()
        {
            isTransaction = false;
            Transaction.Rollback();

        }
        public void CommitTransaction()
        {
            Transaction.Commit();
            isTransaction = false;
        }

        /// <summary>
        /// Establece para el ámbito de conexión actual hacia la base de
        /// datos, se utilizará un manejo explícito de la transacción 
        /// asociada a las consultas, estableciendo el nivel de aislamiento
        /// y bloqueo de recursos de la BD a lo largo durante el transcurso
        /// de la transacción.
        /// </summary>
        /// <param name="level">Nivel de bloqueo de los recursos utilizados.</param>
        public void BeginTransaction(IsolationLevel level = IsolationLevel.ReadCommitted)
        {
            if (!isTransaction && _connection.State == ConnectionState.Open)
            {
                Transaction = _connection.BeginTransaction(level);
                isTransaction = true;
            }
        }

        /// <summary>
        /// Abre el objeto de conexión para realizar las consultas, si es 
        /// necesario o crea un nuevo ámbito de conexión.
        /// </summary>
        public DbCommand FillCommand(string sqlCommand, Object[] values, Object[] paramDirs = null)
        {
            this.LoadCommandObj(sqlCommand, values, paramDirs);

            return Command;
        }

        /// <summary>
        /// Carga los parámetros de un procedimiento almacenado a la BD, sin hacer ninguna ejecución contra el objeto de conexión.
        /// </summary>
        /// <param name="procedureName">Nombre del procedimiento.</param>
        /// <param name="values">Valores actuales para los parámetros establecidos, en conjunción con los valores del procedimiento.</param>
        /// <param name="parameterNames">Nombres de los parametros para la consulta (separados por coma [,]).</param>
        /// <param name="paramDirs">Dirección de los parametros suplidos.</param>
        /// <returns>Un Objeto DbCommand con los parametros cargados.</returns>
        public DbCommand FillCommand(string procedureName, Object[] values, String parameterNames, Object[] paramDirs = null)
        {
            this.LoadCommandObj(procedureName, values, paramDirs, true, parameterNames);
            return Command;
        }

        /// <summary>
        /// Carga los parámetros de la consulta sin hacer ninguna ejecución contra la base de datos y no retorna ningún valor.<br />
        /// Este método sirve de base para la parametrización de los datos de todos los demás métodos disponibles de forma 
        /// pública desde el Helper.
        /// </summary>
        /// <param name="sqlCommand">Comando parametrizado a ejecutar en la BD.</param>
        /// <param name="values">Valores actuales para los parámetros establecidos.</param>
        /// <param name="paramDirs">Dirección de los parametros suplidos.</param>
        /// <param name="isProcedure">Especifica si se va a ejecutar un procedimiento almacenado.</param>
        /// <param name="parameters">Nombres de los parametros para la consulta (separados por coma [,]).</param>
        private void LoadCommandObj(String sqlCommand, Object[] values, Object[] paramDirs = null, bool isProcedure = false, string parameters = null)
        {
            Command = new SqlCommand();
            Command.Connection = _connection;
            if (isTransaction) { Command.Transaction = Transaction; }
            Command.CommandText = sqlCommand;
            ArrayList names = new ArrayList();

            if (isProcedure)
            {
                string[] prms = parameters.Split(',').Select(r => r.Trim()).ToArray();
                names.AddRange(prms);
                Command.CommandType = CommandType.StoredProcedure;
            }
            else
            {
                names = GetParameterNames(sqlCommand);
                Command.CommandType = commandType;
            }

            int index = 0;
            if (values != null)
            {
                foreach (Object val in values)
                {
                    if (paramDirs == null)
                    {
                        if (names.Count == 0)
                        {
                            AddWithValue("", val);
                        }
                        else
                        {
                            AddWithValue(names[index].ToString(), val);
                        }

                    }
                    else
                    {
                        if (paramDirs.Contains(names[index]))
                        {
                            AddWithValue(names[index].ToString(), val, ParameterDirection.Output);
                        }
                        else if (!paramDirs.Contains(names[index]) && paramDirs.Length < values.Length)
                        {
                            AddWithValue(names[index].ToString(), val);
                        }
                        else
                        {
                            AddWithValue(
                                names[index].ToString(), val,
                                (ParameterDirection)Enum.Parse(typeof(ParameterDirection),
                                    paramDirs[index].ToString()));
                        }
                    }

                    index++;
                }
            }
        }

        /// <summary>
        /// Método utilitario para obtener los nombres de los parámetros especificados
        /// en una sentencia SQL dada.
        /// </summary>
        /// <typeparam name="T">Tipo asociado al comando que ejecutará la sentencia.</typeparam>
        /// <param name="query">Sentencia SQL a ejecutar.</param>
        /// <returns>ArrayList con los nombres de los parametros.</returns>
        public ArrayList GetParameterNames(string query)
        {
            ArrayList paramNames = new ArrayList();

            Regex pattern = new Regex(":([A-Za-z0-9]+)(\\s*)");
            foreach (Match match in pattern.Matches(query))
            {
                paramNames.Add(match.Value.Replace(':', ' ').Trim());
            }



            return paramNames;
        }

        /// <summary>
        /// Agrega un DbParameter al DbCommand asociado.
        /// </summary>
        /// <param name="comm">
        /// Objeto DbCommand.
        /// </param>
        /// <param name="paramName">
        /// El nombre del parámetro.
        /// </param>
        /// <param name="value">
        /// El valor del parámetro.
        /// </param>
        /// <remarks>
        /// </remarks>
        public void AddWithValue
        (
            string paramName,
            object value,
            ParameterDirection direction = ParameterDirection.Input
        )
        {
            var param = Command.CreateParameter();
            param.ParameterName = paramName;
            param.Value = value;
            param.Direction = direction;
            Command.Parameters.Add(param);
        }

        public DataSet ExecuteCursor(String dsName = "DataTable")
        {
            _da = new SqlDataAdapter();
            _da.TableMappings.Add("Table", dsName);
            DataSet ds = new DataSet("DataTable");
            _da.SelectCommand = Command;
            _da.Fill(ds);

            return ds;
        }

        //public DataSet ReturnDsFromCursor(string procedureName, Object[] values, String parameterNames, Object[] paramDirs = null, String parameterOut = "")
        //{
        //    //OpenConnection();
        //    //this.LoadCommandObj(procedureName, values, paramDirs, true, parameterNames);
        //    ////return Command;
        //    //Command.Parameters[parameterOut].SqlDbType = SqlDbType..RefCursor;
        //    //Ds = ExecuteCursor();
        //    //CloseConnection();
        //    //return Ds;
        //}

        /// <summary>
        /// Ejecuta una sentencia SQL que devuelve un valor único, usando el objeto de conexión, el comando
        /// establecido como parámetro, y los valores suplidos.
        /// </summary>
        /// <param name="sqlCommand">Comando parametrizado a ejecutar en la BD.</param>
        /// <param name="values">Valores actuales para los parámetros establecidos.</param>
        /// <param name="paramDirs">Dirección de los parametros suplidos.</param>
        /// <returns>Objeto con el primer valor de la primera columna del resultado de la consulta.</returns>
        public Object ExecuteScalar(string sqlCommand, Object[] values, Object[] paramDirs = null)
        {
            Object returnVal = null;
            this.LoadCommandObj(sqlCommand, values, paramDirs);

            returnVal = Command.ExecuteScalar();

            return returnVal;
        }

        public DataSet ReturnDsFromCursor(string procedureName, Object[] values, String parameterNames, Object[] paramDirs = null, String parameterOut = "")
        {
            OpenConnection();
            this.LoadCommandObj(procedureName, values, paramDirs, true, parameterNames);
            //return Command;
            //Command.Parameters[parameterOut].SqlDbType = SqlDbType.RefCursor;
            Ds = ExecuteCursor();
            CloseConnection();
            return Ds;
        }

        /// <summary>
        /// Ejecuta un procedimiento almacenado que devuelve un valor único, usando el objeto de conexión,
        /// el comando establecido como parámetro, y los valores suplidos.
        /// </summary>
        /// <param name="procedureName">Nombre del procedimiento.</param>
        /// <param name="values">Valores actuales para los parámetros establecidos, en conjunción con los valores del procedimiento.</param>
        /// <param name="parameterNames">Nombres de los parametros para la consulta (separados por coma [,]).</param>
        /// <param name="paramDirs">Dirección de los parametros suplidos.</param>
        /// <returns>Objeto con el primer valor de la primera columna del resultado de la consulta.</returns>
        public Object ExecuteScalar(string procedureName, Object[] values, String parameterNames, Object[] paramDirs = null)
        {
            Object returnVal = null;
            this.LoadCommandObj(procedureName, values, paramDirs, true, parameterNames);
            returnVal = Command.ExecuteScalar();

            return returnVal;
        }
    }
}
